### DISCUSSION TITLE
HTTPS for selfhosted vaultwarden in the local network

### DISCUSSION THREAD

**Discussion by dominikheinz** [dani-garcia/vaultwarden#2096] [Q&A]
Hello everyone,

I need some clarifications on how I can come up with a secure setup for vaultwarden.
Here is what I have in mind:

I plan on running VW on a raspberry pi, that is only accessible in my local network. For external access/sync I have a VPN connection into my home network. So once I am connected to my VPN, I will essentially be able to connect to my VW instance from anywhere.

Now, the only thing that bugs me is the fact that VW out of the box uses port 80 (http).
But, of course I would like to use HTTPS instead, to prevent MITM attacks. Is it really required to use a domain for this, since essentially the raspberry pi is only running in my local network and not on an external server?
Is it enough if I just use my domain and point the A record to my local ip? Is there anything in particular I have to pay attention to?

What setup would you guys recommend for this? I would like to go with docker-compose, but I am a bit unsure about creating self-signed certificates, and weather I require a domain or not. Some clarifications would be nice.

**Comment by BlackDex** [comment]
https://github.com/dani-garcia/vaultwarden/wiki/Running-a-private-vaultwarden-instance-with-Let%27s-Encrypt-certs

**Reply by anarchy89** [reply]
If I want to access the server using the Bitwarden app on iOS is there a way to do it with this method?

**Reply by pete1019** [reply]
Working fine in my comment here:
https://github.com/dani-garcia/vaultwarden/discussions/2096#discussioncomment-4554191

**Comment by ErdemDegirmenci** [comment]
I want exactly the same thing Dominik. How did you progress on this?

**Reply by pete1019** [reply]
Easy alternative without open ports:
https://github.com/dani-garcia/vaultwarden/discussions/2096#discussioncomment-4510239

**Reply by ErdemDegirmenci** [reply]
On my Home Network;
I use pfSense for firewall, NPM for reverse proxy, AdGuard and Pihole for DNS for different purposes.
My domain's DNS is on Cloudflare. I can access my local resources with Cludflare Tunnel.

My concern is that in this case you have access to vault.mydomain.com from the entire internet. I wanted to restrict access to certain IPs via Cloudflare or pfsense etc., but for various reasons I could not succeed yet.

**Reply by pete1019** [reply]
> On my Home Network; I use pfSense for firewall, NPM for reverse proxy, AdGuard and Pihole for DNS for different purposes. My domain's DNS is on Cloudflare. I can access my local resources with Cludflare Tunnel.
> 
> My concern is that in this case you have access to vault.mydomain.com from the entire internet. I wanted to restrict access to certain IPs via Cloudflare or pfsense etc., but for various reasons I could not succeed yet.

DON'T set up a DNS entry (No A record) for your domain on Cloudflare. This way, your domain will not resolve in the internet. Just in your local LAN. This works just fine for us.

**Comment by pete1019** [comment]
Easy LAN-only alternative without open ports:

- nginx proxy manager
- free Cloudflare account
- domain
- dns challenge to cloudflare (API)
- pihole for DNS and redirect your domain to nginx proxy manager

See: https://youtu.be/TBGOJA27m_0

VPN like Wireguard needed with set DNS to Pihole needed when accessing from outside of LAN.

**Reply by fedonr** [reply]
Can you explain that in more detail? I am trying to do the same, I'm wondering if I can use DuckDNS or noIP domain or if there is any other Free domain I can use, and I'm also wondering how to do the last step mentioned above about PiHole and if that possible to do the same using Adguard ?

**Reply by pete1019** [reply]
Pihole or Adguard does not matter. Both are for DNS.
When you redirect sudomain.mydomain.com in Adguard or Pihole to your Nginx Proxy Manager like 192.168.1.10 it will stay all internally.
Idea: your domain will be resolved internally.

Just try and tell if you get stuck and give as much information as possible.

**Reply by fedonr** [reply]
THank you so are you referring me to get a actual DOmain or I can use the Adguard's DNS rewrites to create a custom domain for internal use? Sorry for so many question, very new to all this

**Reply by pete1019** [reply]
> THank you so are you referring me to get a actual DOmain or I can use the Adguard's DNS rewrites to create a custom domain for internal use? Sorry for so many question, very new to all this

You can try a dyndns which is a domain as well.
Add this dyndns domain to your Adguard to forward to nginx proxy manager.
Don't forget to setup let's encrypt certificate with nginx proxy manager.

Make sure to really use Adguard as DNS in your router.

**Reply by pete1019** [reply]
Check this:
https://youtu.be/yMmxw-DZ5Ec

**Reply by fedonr** [reply]
Thanks a lot for the help brother, I did check it out. But while I asked for the first time here, I was halfway on using Caddy as per the documentation and I was stuck so I came here and asked around. But fortune caddy thing worked out for me and Vaultwarden works great. Thanks a lot to you @pete1019 and this community for such amazing help.

**Reply by pete1019** [reply]
> Thanks a lot for the help brother, I did check it out. But while I asked for the first time here, I was halfway on using Caddy as per the documentation and I was stuck so I came here and asked around. But fortune caddy thing worked out for me and Vaultwarden works great. Thanks a lot to you @pete1019 and this community for such amazing help.

Please share your caddy solution with everybody.

**Reply by fedonr** [reply]
Sure brother. Here it is follow the steps mentioned here and also use his repo with the tweaks shared below : https://youtu.be/pH_LZVfuSWo

You can use his github repo, but rather than pulling it, I'll suggest make your own files, copying the content from his repo, as it requires few tweaks to make it work.

The main tweak required to make it work is in Caddyfile

Instead of using the format tls { dns duckdns {token} }.
You need to remove {} from token and make it : tls {dns duckdns token } to make it work.

Ensure to the Caddyfile includes the correct container name and rocket ports of Vaultwarden. Default is vaultwarden:80

There are 3 main files to use from repo: Caddyfile, .env file and docker-compose file.

But I'll rather skip the .env file and copy those as environment variables in docker-compose itself for caddy. And would also map ports directly rather than using variables from .env file

Update : I also tried the same thing with Passbolt as well.  Followed the Networkchuck's video for basic setup, and added this caddy part on top instead of using treafik as mentioned in that video. And Passbolt works great too. But I'll stick with Vaultwarden as passbolt won't even let you access the vault offline (from mobile apps like Bitwarden does)

**Reply by rama31244** [reply]
I use Traefik reverse proxy and have set up CloudFlare DNS challenges. I'm stuck on the step where I need to make my vaultwarden subdomain resolve to my Traefik proxy. I run adguard home rather than pihole and can do DNS rewrites but this only lets me point to an IP address while Traefik runs on a specific port of my server, how can I point my subdomain to Traefik in this situation?

**Comment by anarchy89** [comment]
Will the Bitwarden app work with this?

On 29 Dec 2022, at 03:10, pete1019 ***@***.***> wrote:

﻿

Whoever reads this, that should help you for local (LAN) vaultwarden:
https://github.com/dani-garcia/vaultwarden/wiki/Using-Docker-Compose#caddy-with-dns-challenge

**Reply by pete1019** [reply]
If you don't want to open ports you would need to connect via a VPN (Wireguard for example).

**Reply by MurphyDoe0815** [reply]
But then you won't be able to access your Bitwarden servers with any add on, neither browser nor mobile phone - only via letsencrypt !

**Reply by pete1019** [reply]
That is what the DNS challenge is for. You will have a working let's encrypt certificate without exposing your ports. We did it yesterday. Working just fine.
Combining free Cloudflare account, a domain and Pihole for DNS will help you:
https://youtu.be/TBGOJA27m_0

Updated my comment here:
https://github.com/dani-garcia/vaultwarden/discussions/2096#discussioncomment-4510239

**Reply by MurphyDoe0815** [reply]
have you tested it with the browser add-on and mobile app? half year ago nothing worked!

**Reply by pete1019** [reply]
@MurphyDoe0815 Yes it is working totally fine with the Mobile app since you will have a 100% valid Lets-Encrypt Cert because of the DNS Challange. Remember that your DNS like Pihole need to forward your domain like bw.mydomain.com to the ip of your NPM. Even when you are connected via VPN. This is easy with Wireguard (DNS variable).

**Reply by rama31244** [reply]
I run an unraid server with docker containers for traefik & vaultwarden and I don't have an individual IP address for traefik since it runs on one of the ports of my unraid server. So when I point bw.mydomain.com to the IP address of the server, it just goes to the unraid login page. How can i make it point to vaultwarden?
